userspace/deepsys/man/cdeepsys.8
.\" Automatically generated by Pandoc 1.19.2.1
.\"
.TH "" "" "" "" ""
.hy
.SS NAME
.PP
cdeepsys \- the ncurses user interface for deepsys
.SS SYNOPSIS
.PP
\f[B]cdeepsys\f[] [\f[I]option\f[]]...
[\f[I]filter\f[]]
.SS DESCRIPTION
.PP
cdeepsys exports deepsys\[aq]s functionality through an intuitive and
powerful ncurses\-based user interface.
.PP
cdeepsys has been designed to mimic tools like \f[B]top\f[] and
\f[B]htop\f[], but it offers richer functionality, including:
.IP \[bu] 2
Support for both live analysis and deepsys trace files.
Trace files can come from the same machine or from another machine.
.IP \[bu] 2
Visibility into a broad range of metrics, including CPU, memory, disk
I/O, network I/O.
.IP \[bu] 2
Ability to observe input/output activity for processes, files, network
connections and more.
.IP \[bu] 2
Ability to drill down into processes, files, network connections and
more to further explore their behavior.
.IP \[bu] 2
Full customization support.
.IP \[bu] 2
Support for deepsys\[aq]s filtering language.
.IP \[bu] 2
Container support by design.
.PP
cdeepsys works on any terminal, and has support for colors and mouse
input.
.PP
\f[B]Views\f[]
.PP
cdeepsys is based on the concept of \[aq]views\[aq], little Lua scripts
that determine how metrics are collected, processed and represented on
screen.
Including a new visualization to cdeepsys doesn\[aq]t require to update
the program, and is simply a matter of adding a new view.
Views rely on the deepsys processing engine, and this means that they can
include any deepsys filter field.
Views are located in the deepsys chisel directory path, usually
\f[I]/usr/share/deepsys/chisels\f[] and \f[I]~/.chisels\f[].
.SS BASIC USAGE
.PP
Here are some basic tips to get you started with deepsys:
.IP "1." 3
If you run cdeepsys without arguments, it will display live system data,
updating every 2 seconds.
To analyze a trace file, use the \-r command line flag.
.IP "2." 3
You can switch to a different view by using the \f[I]F2\f[] key.
.IP "3." 3
You can drill down into a selection by clicking \f[I]enter\f[].
You can navigate back by typing \f[I]backspace\f[].
.IP "4." 3
You can observe input/output for the currently selected entity by typing
\f[I]F5\f[]
.IP "5." 3
You can see deepsys events for the currently selected entity by typing
\f[I]F6\f[]
.SS DRILLING DOWN
.PP
You drill down by selecting an element in a view and then clicking
\f[I]enter\f[].
Once inside a selection, you can switch to a different view, and the new
view will be applied in the context of the selection.
For example, if you drill down into a process called foo and then switch
to the \f[I]Connections\f[] view, the output will include only the
connections made or received by \f[I]foo\f[].
.PP
To drill down multiple times, keep clicking \f[I]enter\f[].
For example, you can click on a container in the \f[I]Containers\f[]
view to get the processes running inside it, and then click on one of
the processes to see its threads.
.SS ACTIONS AND HOTKEYS
.PP
Each view has a list of command lines that can be executed in the
context of the current selection by pressing \[aq]hotkeys\[aq].
For example, pressing \[aq]k\[aq] in the Processes view kills the
selected process, pressing \[aq]b\[aq] in the Containers view opens a
bash shell in the selected container.
.PP
Each view supports different actions.
You can see which actions a view supports by pressing F8.
You can customize the view\[aq]s actions by editing the view\[aq]s Lua
file.
.SS CONTAINERS SUPPORT
.PP
Starting cdeepsys with the \-pc command line switch will cause many of
the views to include additional container information.
For example, the \f[I]Processes\f[] will include a column showing the
container the process belongs to.
Similarly, the \f[I]Connections\f[] view will show which container each
connection belongs to.
.SS INTERACTIVE COMMANDS
.SS Views Window
.PP
\f[B]Arrows, PgUP, PgDn, Home, End\f[]
.PD 0
.P
.PD
Change the selection and scroll view content, both vertically and
horizontally.
.PP
\f[B]Enter\f[]
.PD 0
.P
.PD
Drill down into the currently highlighted entry.
.PP
\f[B]Backspace\f[]
.PD 0
.P
.PD
Navigate back to the previous view.
.PP
\f[B]F2\f[]
.PD 0
.P
.PD
Show the view picker.
This will let you switch to another view.
.PP
\f[B]CTRL+F /\f[]
.PD 0
.P
.PD
Incremental search in the list of view entries.
.PP
\f[B]F4\f[]
.PD 0
.P
.PD
Incremental filtering of the view entries.
.PP
\f[B]F5, e\f[]
.PD 0
.P
.PD
\[aq]echo FDs\[aq] for the selection, i.e.
view FD input/output for the currently highlighted entry.
.PP
\f[B]F6, d\f[]
.PD 0
.P
.PD
\[aq]dig\[aq] into the selection, i.e.
view deepsys events for the currently highlighted entry.
Refer to the deepsys man page to learn about interpreting the content of
this window.
.PP
\f[B]F7\f[]
.PD 0
.P
.PD
Show the help page for the currently displayed view.
.PP
\f[B]F8\f[]
.PD 0
.P
.PD
Open the view\[aq]s actions panel.
.PP
\f[B]F9, >\f[]
.PD 0
.P
.PD
Open the column sort panel.
.PP
\f[B]F10, q\f[]
.PD 0
.P
.PD
Quit.
.PP
\f[B]DEL, c\f[]
.PD 0
.P
.PD
For views that are listing elements without aggregating them by key
(identifiable by yellow column headers), this command clears the view
content.
.PP
\f[B]p\f[]
.PD 0
.P
.PD
Pause screen updates.
.PP
\f[B]\f[C]<shift>\ <1\-9>\f[]\f[]
.PD 0
.P
.PD
sort column \f[C]<n>\f[]
.PP
\f[B]F1, h, ?\f[]
.PD 0
.P
.PD
Show the help screen.
.SS Echo and deepsys Windows
.PP
\f[B]Arrows, PgUP, PgDn, Home, End\f[]
.PD 0
.P
.PD
Scroll the page content.
.PP
\f[B]Backspace\f[]
.PD 0
.P
.PD
Navigate back to the previous view.
.PP
\f[B]CTRL+F /\f[]
.PD 0
.P
.PD
Search inside the window content.
.PP
\f[B]F3\f[]
.PD 0
.P
.PD
Find Next.
.PP
\f[B]F2\f[]
.PD 0
.P
.PD
Chose the output rendering format.
Options are \[aq]Dotted ASCII\[aq] (non\-printable binary bytes are
rendered as dots), \[aq]Printable ASCII\[aq] (non\-printable binary
bytes are not included and line endings are rendered accurately) and
\[aq]Hex\[aq] (dotted ASCII representation is included together with the
Hexadecimal rendering of the buffers).
.PP
\f[B]DEL, c\f[]
.PD 0
.P
.PD
Clear the screen content.
.PP
\f[B]p\f[]
.PD 0
.P
.PD
Pause screen updates.
.PP
\f[B]CTRL+G\f[]
.PD 0
.P
.PD
Go to line.
.SS Spectrogram Window
.PP
\f[B]F2\f[]
.PD 0
.P
.PD
Show the view picker.
This will let you switch to another view.
.PP
\f[B]p\f[]
.PD 0
.P
.PD
Pause/Resume the visualization.
.PP
\f[B]Backspace\f[]
.PD 0
.P
.PD
Navigate back to the previous view.
.SS MOUSE USAGE
.IP \[bu] 2
Clicking on column headers lets you sort the table.
.IP \[bu] 2
Double clicking on row entries performs a drill down.
.IP \[bu] 2
Clicking on the filter string at the top of the screen (the text after
\[aq]Filter:\[aq]) lets you change the deepsys filter and customize the
view content.
.IP \[bu] 2
You can use the mouse on the entries in the menu at the bottom of the
screen to perform their respective actions.
.SS COMMAND LINE OPTIONS
.PP
\f[B]\-d\f[] \f[I]period\f[], \f[B]\-\-delay\f[]=\f[I]period\f[]
.PD 0
.P
.PD
Set the delay between updates, in milliseconds (by default = 2000).
This works similarly to the \-d option in top.
.PP
\f[B]\-E\f[], \f[B]\-\-exclude\-users\f[]
.PD 0
.P
.PD
Don\[aq]t create the user/group tables by querying the OS when deepsys
starts.
This also means that no user or group info will be written to the
tracefile by the \-w flag.
The user/group tables are necessary to use filter fields like user.name
or group.name.
However, creating them can increase deepsys\[aq]s startup time.
.PP
\f[B]\-\-force\-term\-compat\f[]
.PD 0
.P
.PD
Try to configure simple terminal settings (xterm\-1002) that work better
with terminals like putty.
Try to use this flag if you experience terminal issues like the mouse
not working.
.PP
\f[B]\-h\f[], \f[B]\-\-help\f[]
.PD 0
.P
.PD
Print this page
.PP
\f[B]\-k\f[], \f[B]\-\-k8s\-api\f[]
.PD 0
.P
.PD
Enable Kubernetes support by connecting to the API server specified as
argument.
E.g.
"<http://admin:password@127.0.0.1:8080>".
The API server can also be specified via the environment variable
DEEPSYS_K8S_API.
.PP
\f[B]\-K\f[] \f[I]btfile | certfile:keyfile[#password][:cacertfile]\f[],
\f[B]\-\-k8s\-api\-cert=\f[]\f[I]btfile |
certfile:keyfile[#password][:cacertfile]\f[]
.PD 0
.P
.PD
Use the provided files names to authenticate user and (optionally)
verify the K8S API server identity.
Each entry must specify full (absolute, or relative to the current
directory) path to the respective file.
Private key password is optional (needed only if key is password
protected).
CA certificate is optional.
For all files, only PEM file format is supported.
Specifying CA certificate only is obsoleted \- when single entry is
provided for this option, it will be interpreted as the name of a file
containing bearer token.
Note that the format of this command\-line option prohibits use of files
whose names contain \[aq]:\[aq] or \[aq]#\[aq] characters in the file
name.
Option can also be provided via the environment variable

userspace/deepsys/cdeepsys.cpp
/*
Copyright (C) 2013-2014 KhulnaSoft, Ltd.

This file is part of deepsys.

deepsys is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

deepsys is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with deepsys.  If not, see <http://www.gnu.org/licenses/>.
*/

#define __STDC_FORMAT_MACROS

#include <stdio.h>
#include <iostream>
#include <time.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <assert.h>
#include <algorithm>

#include <sinsp.h>
#include "chisel.h"
#include "deepsys.h"
#include "table.h"
#include "utils.h"

#ifdef _WIN32
#include "win32/getopt.h"
#include <io.h>
#else
#include <unistd.h>
#include <getopt.h>
#include <term.h>
#endif

#include "cursescomponents.h"
#include "cursestable.h"
#include "cursesui.h"

#define MOUSE_CAPABLE_TERM "xterm-1003"
#define MOUSE_CAPABLE_TERM_COMPAT "xterm-1002"

static bool g_terminate = false;
static void usage();

//
// Helper functions
//
static void signal_callback(int signal)
{
	g_terminate = true;
}

//
// Program help
//
static void usage()
{
    printf(
"cdeepsys version " DEEPSYS_VERSION "\n"
"Usage: cdeepsys [options] [filter]\n\n"
"Options:\n"
" -A, --print-ascii  When emitting JSON, only print the text portion of data buffers, and echo\n"
"                    end-of-lines. This is useful to only display human-readable\n"
"                    data.\n"
" -d <period>, --delay=<period>\n"
"                    Set the delay between updates, in milliseconds. This works\n"
"                    similarly to the -d option in top.\n"
" -E, --exclude-users\n"
"                    Don't create the user/group tables by querying the OS when\n"
"                    deepsys starts. This also means that no user or group info\n"
"                    will be written to the tracefile by the -w flag.\n"
"                    The user/group tables are necessary to use filter fields\n"
"                    like user.name or group.name. However, creating them can\n"
"                    increase deepsys's startup time. Moreover, they contain\n"
"                    information that could be privacy sensitive.\n"
" --force-term-compat\n"
"                    Try to configure simple terminal settings (xterm-1002) that work\n"
"                    better with terminals like putty. Try to use this flag if you experience\n"
"                    terminal issues like the mouse not working.\n"
" -h, --help         Print this page\n"
" -k <url>, --k8s-api=<url>\n"
"                    Enable Kubernetes support by connecting to the API server\n"
"                    specified as argument. E.g. \"http://admin:password@127.0.0.1:8080\".\n"
"                    The API server can also be specified via the environment variable\n"
"                    DEEPSYS_K8S_API.\n"
" -K <bt_file> | <cert_file>:<key_file[#password]>[:<ca_cert_file>], --k8s-api-cert=<bt_file> | <cert_file>:<key_file[#password]>[:<ca_cert_file>]\n"
"                    Use the provided files names to authenticate user and (optionally) verify the K8S API\n"
"                    server identity.\n"
"                    Each entry must specify full (absolute, or relative to the current directory) path\n"
"                    to the respective file.\n"
"                    Private key password is optional (needed only if key is password protected).\n"
"                    CA certificate is optional. For all files, only PEM file format is supported. \n"
"                    Specifying CA certificate only is obsoleted - when single entry is provided \n"
"                    for this option, it will be interpreted as the name of a file containing bearer token.\n"
"                    Note that the format of this command-line option prohibits use of files whose names contain\n"
"                    ':' or '#' characters in the file name.\n"
"                    Option can also be provided via the environment variable DEEPSYS_K8S_API_CERT.\n"
" -l, --list         List all the fields that can be used in views.\n"
" --logfile=<file>\n"
"                    Print program logs into the given file.\n"
" -n <num>, --numevents=<num>\n"
"                    Stop capturing after <num> events\n"
" --page-faults      Capture user/kernel major/minor page faults\n"
" -pc, -pcontainer\n"
"                    Instruct cdeepsys to use a container-friendly format in its\n"
"                    views.\n"
"                    This will cause several of the views to contain additional\n"
"                    container-related columns.\n"
" -R                 Resolve port numbers to names.\n"
" -r <readfile>, --read=<readfile>\n"
"                    Read the events from <readfile>.\n"
" --raw              Print raw output on a regular terminal instead of enabling\n"
"                    ncurses-based ANSI output.\n"
" -s <len>, --snaplen=<len>\n"
"                    Capture the first <len> bytes of each I/O buffer.\n"
"                    By default, the first 80 bytes are captured. Use this\n"
"                    option with caution, it can generate huge trace files.\n"
" -T, --force-tracers-capture\n"
"                    Tell the driver to make sure full buffers are captured from\n"
"                    /dev/null, to make sure that tracers are completely\n"
"                    captured. Note that deepsys will enable extended /dev/null\n"
"                    capture by itself after detecting that tracers are written\n"
"                    there, but that could result in the truncation of some\n"
"                    tracers at the beginning of the capture. This option allows\n"
"                    preventing that.\n"
" -v <view_id>, --view=<view_id>\n"
"                    Run the view with the given ID when cdeepsys starts.\n"
"                    View IDs can be found in the view documentation pages in\n"
"                    cdeepsys. Combine  this option with a command line filter for\n"
"                    complete output customization.\n"
" --version          Print version number.\n"
" -X, --print-hex-ascii\n"
"                    When emitting JSON, print data buffers in hex and ASCII.\n"
"\n"
"How to use cdeepsys:\n"
"1. you can either see real time data, or analyze a trace file by using the -r\n"
"   command line flag.\n"
"2. you can switch to a different view by using the F2 key.\n"
"3. You can drill down into a selection by typing enter.\n"
"   You can navigate back by typing backspace.\n"
"4. you can observe reads and writes (F5) or see deepsys events (F6) for any\n"
"   selection.\n"
"\nAdditional help can be obtained by clicking F1 while the program is running,\n"
"and in the man page.\n\n"
    );
}

#ifdef HAS_CHISELS
static void add_chisel_dirs(sinsp* inspector)
{
	//
	// Add the default chisel directory statically configured by the build system
	//
	inspector->add_chisel_dir(DEEPSYS_INSTALLATION_DIR CHISELS_INSTALLATION_DIR, false);

	//
	// Add the directories configured in the DEEPSYS_CHISEL_DIR environment variable
	//
	char* s_user_cdirs = getenv("DEEPSYS_CHISEL_DIR");

	if(s_user_cdirs != NULL)
	{
		vector<string> user_cdirs = sinsp_split(s_user_cdirs, ';');

		for(uint32_t j = 0; j < user_cdirs.size(); j++)
		{
			inspector->add_chisel_dir(user_cdirs[j], true);
		}
	}
}

static void print_views(sinsp_view_manager* view_manager)
{
	Json::FastWriter writer;
	Json::Value root;

	vector<sinsp_view_info>* vlist = view_manager->get_views();

	for(auto it = vlist->begin(); it != vlist->end(); ++it)
	{
		Json::Value jv;
		sinsp_view_info& vinfo = *it;

		jv["id"] = vinfo.m_id;
		jv["name"] = vinfo.m_name;
		jv["description"] = vinfo.m_description;
		jv["isRoot"] = vinfo.m_is_root;
		jv["drilldownTarget"] = vinfo.m_drilldown_target;
		jv["filter"] = vinfo.m_filter;
		jv["canDrillDown"] = (vinfo.m_type == sinsp_view_info::T_TABLE);

		for(auto it = vinfo.m_applies_to.begin(); it != vinfo.m_applies_to.end(); ++it)
		{
			jv["appliesTo"].append(*it);
		}
		for(auto it = vinfo.m_tags.begin(); it != vinfo.m_tags.end(); ++it)
		{
			jv["tags"].append(*it);
		}
		for(auto it = vinfo.m_tips.begin(); it != vinfo.m_tips.end(); ++it)
		{
			jv["tips"].append(*it);
		}

		root.append(jv);
	}

	string ouput = writer.write(root);
	printf("%s", ouput.substr(0, ouput.size() - 1).c_str());
}
#endif

captureinfo do_inspect(sinsp* inspector,
					   uint64_t cnt,
					   sinsp_cursesui* ui)
{
	captureinfo retval;
	int32_t res;
	sinsp_evt* ev;

	//
	// Loop through the events
	//
	while(1)
	{
		if(retval.m_nevts == cnt || g_terminate)
		{
			//
			// End of capture, either because the user stopped it, or because
			// we reached the event count specified with -n.
			//
			break;
		}

		res = inspector->next(&ev);

		if(res == SCAP_TIMEOUT)
		{
			continue;
		}
		else if(res != SCAP_EOF && res != SCAP_SUCCESS)
		{
			//
			// Event read error.
			// Notify the chisels that we're exiting, and then die with an error.
			//
			if(inspector->is_live())
			{
				throw sinsp_exception(inspector->getlasterr());
			}
			else
			{
				ui->set_truncated_input(true);
				res = SCAP_EOF;
				continue;
			}
		}

		if(ui->process_event(ev, res) == true)
		{
			return retval;
		}

		retval.m_nevts++;
	}

	return retval;
}

string g_version_string = DEEPSYS_VERSION;

deepsys_init_res cdeepsys_init(int argc, char **argv)
{
	deepsys_init_res res;
	sinsp* inspector = NULL;
	vector<string> infiles;
	int op;
	uint64_t cnt = -1;
	uint32_t snaplen = 0;
	int long_index = 0;
	int32_t n_filterargs = 0;
	captureinfo cinfo;
	string errorstr;
	string display_view;
	bool print_containers = false;
	uint64_t refresh_interval_ns = 2000000000;
	bool list_flds = false;
	bool is_interactive = false;
	int32_t json_first_row = 0;
	int32_t json_last_row = 0;
	int32_t sorting_col = -1;
	bool list_views = false;

#ifndef _WIN32
	sinsp_table::output_type output_type = sinsp_table::OT_CURSES;
#else
	sinsp_table::output_type output_type = sinsp_table::OT_JSON;
#endif
	string* k8s_api = 0;
	string* k8s_api_cert = 0;
	string* mesos_api = 0;
	bool terminal_with_mouse = false;
	bool force_tracers_capture = false;
	bool force_term_compat = false;
	sinsp_evt::param_fmt event_buffer_format = sinsp_evt::PF_NORMAL;
	bool page_faults = false;
	
	static struct option long_options[] =
	{
		{"print-ascii", no_argument, 0, 'A' },
		{"delay", required_argument, 0, 'd' },
		{"exclude-users", no_argument, 0, 'E' },
		{"from", required_argument, 0, 0 },
		{"help", no_argument, 0, 'h' },
		{"k8s-api", required_argument, 0, 'k'},
		{"k8s-api-cert", required_argument, 0, 'K' },
		{"json", no_argument, 0, 'j' },
		{"interactive", optional_argument, 0, 0 },
		{"list", optional_argument, 0, 'l' },
		{"list-views", no_argument, 0, 0},
		{"mesos-api", required_argument, 0, 'm'},
		{"numevents", required_argument, 0, 'n' },
		{"page-faults", no_argument, 0, 0 },
		{"print", required_argument, 0, 'p' },
		{"resolve-ports", no_argument, 0, 'R'},
		{"readfile", required_argument, 0, 'r' },
		{"raw", no_argument, 0, 0 },
		{"snaplen", required_argument, 0, 's' },
		{"logfile", required_argument, 0, 0 },
		{"force-tracers-capture", required_argument, 0, 'T'},
		{"force-term-compat", no_argument, 0, 0},
		{"sortingcol", required_argument, 0, 0 },
		{"to", required_argument, 0, 0 },
		{"view", required_argument, 0, 'v' },
		{"version", no_argument, 0, 0 },
		{"print-hex-ascii", no_argument, 0, 'X'},
		{0, 0, 0, 0}
	};

	//
	// Parse the arguments
	//
	try
	{
		inspector = new sinsp();

#ifdef HAS_CHISELS
		add_chisel_dirs(inspector);
#endif

		//
		// Parse the args
		//
		while((op = getopt_long(argc, argv,
			"Ad:Ehk:K:jlm:n:p:Rr:s:Tv:X", long_options, &long_index)) != -1)
		{
			switch(op)
			{
			case '?':
				//
				// Command line error 
				//
				throw sinsp_exception("command line error");
				break;
			case 'A':
				if(event_buffer_format != sinsp_evt::PF_NORMAL)
				{
					fprintf(stderr, "you cannot specify more than one output format\n");
					delete inspector;
					return deepsys_init_res(EXIT_SUCCESS);
				}

				event_buffer_format = sinsp_evt::PF_EOLS_COMPACT;
				break;
			case 'd':
				try
				{
					refresh_interval_ns = sinsp_numparser::parseu64(optarg) * 1000000;
				}
				catch(...)
				{
					throw sinsp_exception("can't parse the -d argument, make sure it's a number");
				}

				if(refresh_interval_ns < 100000000)
				{
					throw sinsp_exception("Period must be bigger then 100ms");
				}

				break;
			case 'E':
				inspector->set_import_users(false);
				break;
			case 'h':
				usage();
				delete inspector;
				return deepsys_init_res(EXIT_SUCCESS);
			case 'k':
				k8s_api = new string(optarg);
				break;
			case 'K':
				k8s_api_cert = new string(optarg);
				break;
			case 'j':
				output_type = sinsp_table::OT_JSON;
				break;
			case 'l':
				list_flds = true;
				break;
			case 'm':
				mesos_api = new string(optarg);
				break;
			case 'n':
				try
				{
					cnt = sinsp_numparser::parseu64(optarg);
				}
				catch(...)
				{
					throw sinsp_exception("can't parse the -n argument, make sure it's a number");
				}

				if(cnt <= 0)
				{
					throw sinsp_exception(string("invalid event count ") + optarg);
					res.m_res = EXIT_FAILURE;
					goto exit;
				}
				break;
			case 'p':
				if(string(optarg) == "c" || string(optarg) == "container")
				{
					inspector->set_print_container_data(true);
					print_containers = true;
				}

				break;
			case 'R':
				inspector->set_hostname_and_port_resolution_mode(true);
				break;
			case 'r':
				infiles.push_back(optarg);
				k8s_api = new string();
				mesos_api = new string();
				break;
			case 's':
				snaplen = atoi(optarg);
				break;
			case 'T':
				force_tracers_capture = true;
				break;
			case 'v':
				display_view = optarg;
				break;
			case 'X':
				if(event_buffer_format != sinsp_evt::PF_NORMAL)
				{
					fprintf(stderr, "you cannot specify more than one output format\n");
					delete inspector;
					return deepsys_init_res(EXIT_FAILURE);
				}

				event_buffer_format = sinsp_evt::PF_HEXASCII;
				break;
			case 0:
				{
					if(long_options[long_index].flag != 0)
					{
						break;
					}

					string optname = string(long_options[long_index].name);
					if(optname == "version")
					{
						printf("deepsys version %s\n", DEEPSYS_VERSION);
						delete inspector;
						return deepsys_init_res(EXIT_SUCCESS);
					}
					else if(optname == "interactive")
					{
						is_interactive = true;
						output_type = sinsp_table::OT_JSON;
					}
					else if(optname == "logfile")
					{
						inspector->set_log_file(optarg);
					}
					else if(optname == "raw")
					{
						output_type = sinsp_table::OT_RAW;
					}
					else if(optname == "force-term-compat")
					{
						force_term_compat = true;
					}
					else if(optname == "from")
					{
						json_first_row = sinsp_numparser::parsed32(optarg);
					}
					else if(optname == "to")
					{
						json_last_row = sinsp_numparser::parsed32(optarg);
					}
					else if(optname == "sortingcol")
					{
						sorting_col = sinsp_numparser::parsed32(optarg);
					}
					else if(optname == "list-views")
					{
						list_views = true;
					}
					else if(optname == "page-faults")
					{
						page_faults = true;
					}
				}
				break;
			default:
				break;
			}
		}

		string filter;

		//
		// If -l was specified, print the fields and exit
		//
		if(list_flds)
		{
			list_fields(false, false);

			res.m_res = EXIT_SUCCESS;
			goto exit;
		}

		//
		// the filter is at the end of the command line
		//
		if(optind + n_filterargs < argc)
		{
#ifdef HAS_FILTERING
			for(int32_t j = optind + n_filterargs; j < argc; j++)
			{
				filter += argv[j];
				if(j < argc)
				{
					filter += " ";
				}
			}
#else
			fprintf(stderr, "filtering not compiled.\n");
			res.m_res = EXIT_FAILURE;
			goto exit;
#endif
		}

		if(signal(SIGINT, signal_callback) == SIG_ERR)
		{
			fprintf(stderr, "An error occurred while setting SIGINT signal handler.\n");
			res.m_res = EXIT_FAILURE;
			goto exit;
		}

		if(signal(SIGTERM, signal_callback) == SIG_ERR)
		{
			fprintf(stderr, "An error occurred while setting SIGTERM signal handler.\n");
			res.m_res = EXIT_FAILURE;
			goto exit;
		}

		if(json_last_row < json_first_row)
		{
			fprintf(stderr, "'to' argument cannot be smaller than the 'from' one.\n");
			res.m_res = EXIT_FAILURE;
			goto exit;
		}

		//
		// Initialize ncurses
		//
#ifndef NOCURSESUI
		if(output_type == sinsp_table::OT_CURSES)
		{
			//
			// Check if terminal has mouse support
			//
			const char* mct = force_term_compat? MOUSE_CAPABLE_TERM_COMPAT : MOUSE_CAPABLE_TERM;
			terminal_with_mouse = (tgetent(NULL, mct) != 0);

			if(terminal_with_mouse)
			{
				//
				// Enable fine-grained mouse activity capture by setting xterm-1002
				//
				setenv("TERM", mct, 1);
			}

			(void) initscr();      // initialize the curses library
			(void) nonl();         // tell curses not to do NL->CR/NL on output
			intrflush(stdscr, false);
			keypad(stdscr, true);
			curs_set(0);
			if(has_colors())
			{
			  start_color();
			}
			use_default_colors();
			mousemask(ALL_MOUSE_EVENTS | REPORT_MOUSE_POSITION, NULL);
			noecho();

			timeout(0);

			// If this is uncommented, it's possible to natively handle stuff like CTRL+c
			//raw();
		}
#endif

		//
		// Create the list of views
		//
		sinsp_view_manager view_manager;

		//
		// Scan the chisel list to load the Lua views, and add them to the list
		//
		vector<chisel_desc> chlist;
		sinsp_chisel::get_chisel_list(&chlist);

		for(auto it : chlist)
		{
			if(it.m_viewinfo.m_valid)
			{
				if(print_containers)
				{
					it.m_viewinfo.apply_tag("containers");
				}
				else
				{
					it.m_viewinfo.apply_tag("default");
				}

				if(it.m_viewinfo.m_tags.size() != 0)
				{
					if(it.m_viewinfo.m_tags[0] == "Containers")
					{
						continue;
					}
				}

				if(output_type != sinsp_table::OT_JSON)
				{
					if(std::find(it.m_viewinfo.m_tags.begin(), 
						it.m_viewinfo.m_tags.end(), 
						"nocdeepsys") != it.m_viewinfo.m_tags.end())
					{
						continue;
					}
				}

				view_manager.add(&it.m_viewinfo);
			}
		}

		//
		// Set the initial disply view
		//
		view_manager.set_selected_view(display_view);

		if(list_views)
		{
			print_views(&view_manager);
			goto exit;
		}

		//
		// Go through the input sources and apply the processing to all of them
		//
		for(uint32_t j = 0; j < infiles.size() || infiles.size() == 0; j++)
		{
			//
			// Initialize the UI
			//
			sinsp_cursesui ui(inspector, 
				(infiles.size() != 0)? infiles[0] : "",
				(filter.size() != 0)? filter : "",
				refresh_interval_ns,
				print_containers,
				output_type,
				terminal_with_mouse,
				json_first_row, 
				json_last_row,
				sorting_col,
				event_buffer_format);

			ui.configure(&view_manager);

			if(display_view == "dig" || display_view == "echo")
			{
				ui.start(false, true);
			}
			else
			{
				ui.start(false, false);
			}

			if(is_interactive)
			{
				printf("ready\n");

				//
				// In interactive mode, make sure stderr is flushed at every printf
				//
				setbuf(stderr, NULL);

				//
				// Set the UI in interactive mode and start listening to user
				// input.
				//
				ui.set_interactive(true);
			}

			//
			// Launch the capture
			//
			if(infiles.size() != 0)
			{
				//
				// We have a file to open
				//
				inspector->open(infiles[j]);
			}
			else
			{
				if(j > 0)
				{
					break;
				}

				//
				// No file to open, this is a live capture
				//
#if defined(HAS_CAPTURE)
				bool open_success = true;
				
				try
				{
					inspector->open("");
				}
				catch(sinsp_exception e)
				{
					open_success = false;
				}

				//
				// Starting the live capture failed, try to load the driver with
				// modprobe.
				//
				if(!open_success)
				{
					open_success = true;

					if(system("modprobe " PROBE_NAME " > /dev/null 2> /dev/null"))
					{
						fprintf(stderr, "Unable to load the driver\n");
					}

					inspector->open("");
				}
#else
				//
				// Starting live capture
				// If this fails on Windows and OSX, don't try with any driver
				//
				inspector->open("");
#endif

				//
				// Enable gathering the CPU from the kernel module
				//
				inspector->set_get_procs_cpu_from_driver(true);
			}

			//
			// If required, set the snaplen
			//
			if(snaplen != 0)
			{
				inspector->set_snaplen(snaplen);
			}

			//
			// If required, tell the driver to enable tracers capture
			//
			if(force_tracers_capture)
			{
				inspector->enable_tracers_capture();
			}

			if(page_faults)
			{
				inspector->enable_page_faults();
			}

			//
			// run k8s, if required
			//
			if(k8s_api)
			{
				if(!k8s_api_cert)
				{
					if(char* k8s_cert_env = getenv("DEEPSYS_K8S_API_CERT"))
					{
						k8s_api_cert = new string(k8s_cert_env);
					}
				}
				inspector->init_k8s_client(k8s_api, k8s_api_cert);
				k8s_api = 0;
				k8s_api_cert = 0;
			}
			else if(char* k8s_api_env = getenv("DEEPSYS_K8S_API"))
			{
				if(k8s_api_env != NULL)
				{
					if(!k8s_api_cert)
					{
						if(char* k8s_cert_env = getenv("DEEPSYS_K8S_API_CERT"))
						{
							k8s_api_cert = new string(k8s_cert_env);
						}
					}
					k8s_api = new string(k8s_api_env);
					inspector->init_k8s_client(k8s_api, k8s_api_cert);
				}
				else
				{
					delete k8s_api;
					delete k8s_api_cert;
				}
				k8s_api = 0;
				k8s_api_cert = 0;
			}

			//
			// run mesos, if required
			//
			if(mesos_api)
			{
				inspector->init_mesos_client(mesos_api);
			}
			else if(char* mesos_api_env = getenv("DEEPSYS_MESOS_API"))
			{
				if(mesos_api_env != NULL)
				{
					mesos_api = new string(mesos_api_env);
					inspector->init_mesos_client(mesos_api);
				}
			}
			delete mesos_api;
			mesos_api = 0;

			if(output_type == sinsp_table::OT_JSON)
			{
				printf("{\"slices\": [\n");
				if(display_view != "dig" && display_view != "echo")
				{
					printf("{\"progress\": 0},\n");
				}
			}

			//
			// Start the capture loop
			//
			cinfo = do_inspect(inspector,
				cnt,
				&ui);

			if(output_type == sinsp_table::OT_JSON)
			{
				printf("]}\n");
				//printf("%c", EOF);
			}

			//
			// Done. Close the capture.
			//
			inspector->close();
		}
	}
	catch(sinsp_capture_interrupt_exception&)
	{
	}
	catch(std::exception& e)
	{
		errorstr = e.what();
		res.m_res = EXIT_FAILURE;
	}
	catch(...)
	{
		errorstr = "uncaught exception";
		res.m_res = EXIT_FAILURE;
	}

exit:
	if(inspector)
	{
		delete inspector;
	}

	//
	// Restore the original screen
	//
#ifndef NOCURSESUI
	if(output_type == sinsp_table::OT_CURSES)
	{
		endwin();
	}
#endif

	if(errorstr != "")
	{
		cerr << errorstr << endl;
	}

	return res;
}

//
// MAIN
//
int main(int argc, char **argv)
{
	deepsys_init_res res;

	//
	// Run cdeepsys
	//
	res = cdeepsys_init(argc, argv);

#ifdef _WIN32
	_CrtDumpMemoryLeaks();
#endif

	return res.m_res;
}
